src/metabase/api/dashboard.clj

```diff
@@ -784,38 +786,51 @@
   "How many results to return when chain filtering"
   1000)
 
-(mu/defn ^:private mappings->field-ids :- [:maybe [:set ms/PositiveInt]]
-  [parameter-mappings :- [:maybe [:or [:set dashboard-card/ParamMapping] [:sequential dashboard-card/ParamMapping]]]]
-  (set (for [{{:keys [card]} :dashcard :keys [target]} parameter-mappings
-             :let  [field-clause (params/param-target->field-clause target card)]
-             :when field-clause
-             :let  [{:keys [result_metadata]} card
-                    field-id (or
-                              ;; Get the field id from the field-clause if it contains it. This is the common case for
-                              ;; mbql queries.
-                              (mbql.u/match-one field-clause [:field (id :guard integer?) _] id)
-                              ;; Attempt to get the field clause from the model metadata corresponding to the field.
-                              ;; This is the common case for native queries in which mappings from original columns
-                              ;; have been performed using model metadata.
-                              (:id (qp.util/field->field-info field-clause result_metadata)))]
-             :when field-id]
-         field-id)))
-
-(defn- param-key->field-ids
-  "Get Field ID(s) associated with a parameter in a Dashboard.
-
-    (param-key->field-ids (t2/select-one Dashboard :id 62) \"ee876336\")
-    ;; -> #{276}"
-  [dashboard param-key]
-  {:pre [(string? param-key)]}
-  (let [{:keys [resolved-params]} (t2/hydrate dashboard :resolved-params)
-        param                     (get resolved-params param-key)]
-    (mappings->field-ids (:mappings param))))
-
-(defn- chain-filter-constraints [dashboard constraint-param-key->value]
-  (into {} (for [[param-key value] constraint-param-key->value
-                 field-id          (param-key->field-ids dashboard param-key)]
-             [field-id value])))
+(defn- find-template-tag
+  "Fetch the `:field` clause from `dashcard` referenced by `template-tag`.
+
+    (find-template-tag [:template-tag :company] some-dashcard) ; -> [:field 100 nil]"
+  [dimension card]
+  (when-let [tag (first (mbql.u/get-clause :template-tag dimension))]
+    (get-in card [:dataset_query :native :template-tags (u/qualified-name tag)])))
+
+(defn- param->op [{:keys [type]}]
+  (if (get-in mbql.s/parameter-types [type :operator])
+    (keyword (name type))
+    :=))
+
+(mu/defn ^:private param->fields
+  [{:keys [mappings] :as param} :- mbql.s/Parameter]
+  (for [{:keys [target] {:keys [card]} :dashcard} mappings
+        :let  [dimension (->> (mbql.normalize/normalize-tokens target :ignore-path)
+                              (mbql.u/get-clause :dimension)
+                              first)]
+        :when dimension
+        :let  [ttag      (find-template-tag dimension card)
+               dimension (if ttag
+                           (:dimension ttag)
+                           dimension)
```

Comment:
is there a clearer way to write this part? the switching back and forth between `dimension` and tag makes it confusing.

in case `ttag` is nil, we returned `dimension` as `dimension` from the previous state on line 808.
It's not clear to me why we do this.

Reply:
Well, if dimension is resolved to a `:template-tag`, we want to get the dimension from that tag. But if not, then it could be a field already (if I'm understanding [tests](https://github.com/metabase/metabase/blob/34039-fix-is-not-filter/test/metabase/api/dashboard_test.clj#L485) correctly).

Reply:
hmmm, yea but it's still very confusing when you read this, like it's hard understand "why" we need to do that.

maybe we need to rename `find-template-tag` to `get-template-tag-if-exists` to make it more explicit, and can you add the explanation above as a comment?

Reply:
How do you feel about `maybe-template-tag`? 😁

Reply:
I wonder if it's better to dispatch on its type, like 
```
(cond 
  (mbql.u/is-clause? :template-tag dimension) (:dimension ttag)
  (mbq.u/is-clause? :field dimension) dimension
  :else (throw (ex-info "no idea" {:dimension dimension}))))
```

Reply:
that's definitely better

Reply:
done in 39ec5a256ca0cd926f3cbad9232029374e1191fe